lahja/asyncio/endpoint.py

```diff
@@ -360,16 +382,77 @@ def run(self, *args, **kwargs):  # type: ignore
     async def start(self) -> None:
         if self.is_running:
             raise RuntimeError(f"Endpoint {self.name} is already running")
+
         self._receiving_loop_running = asyncio.Event()
         self._receiving_queue = asyncio.Queue()
 
+        self._subscription_updates_running = asyncio.Event()
+        self._subscription_updates_condition = asyncio.Condition()
+        self._subscription_updates_queue = asyncio.Queue()
+
         self._running = True
 
         self._endpoint_tasks.add(asyncio.ensure_future(self._connect_receiving_queue()))
+        self._endpoint_tasks.add(
+            asyncio.ensure_future(self._process_subscription_updates())
+        )
+        self._endpoint_tasks.add(
+            asyncio.ensure_future(self._process_subscription_updates_queue())
+        )
 
         await self._receiving_loop_running.wait()
+        await self._subscription_updates_running.wait()
+
         self.logger.debug("Endpoint[%s]: running", self.name)
 
+    async def _process_subscription_updates_queue(self) -> None:
+        while self.is_running:
+            await self._subscription_updates_queue.get()
+            async with self._subscription_updates_condition:
+                self._subscription_updates_condition.notify_all()
+
+    async def _process_subscription_updates(self) -> None:
+        self._subscription_updates_running.set()
+        async with self._subscription_updates_condition:
+            while self.is_running:
+                await self._subscription_updates_condition.wait()
+                await self._do_notify_subscriptions_changed()
+
+    def _notify_subscriptions_changed_nowait(self) -> None:
+        self._subscription_updates_queue.put_nowait(None)
+
+    async def _do_notify_subscriptions_changed(self) -> None:
+        """
+        Tell all inbound connections of our new subscriptions
+        """
+        # make a copy so that the set doesn't change while we iterate over it
+        subscribed_events = self.subscribed_events
+        await asyncio.gather(
+            *(
+                remote.notify_subscriptions_updated(subscribed_events)
+                for remote in itertools.chain(
+                    self._half_connections.copy(), self._full_connections.values()
+                )
+            )
+        )
+
+    async def _connect_receiving_queue(self) -> None:
```

Comment:
Didn't intend to relocate this but I think I like it here better.  Need to add some comments to deliniate the body of this class more clearly but I'll do that as a cleanup PR since re-organizing makes code hard to audit.

TODO: revert this move.